aptos-labs/aptos-core
# [CLI][Framework][Ledger] Add Ledger key rotation and more profile management to CLI, fix/extend associated Framework code

@banool @davidiw @gedigi @gregnazario @hardsetting

## Summary of changes

This PR has been resumed after several months of delays, since the Ledger C APIs
needed to be extended.

This PR pairs with corresponding tutorial docs at
https://github.com/aptos-labs/developer-docs/pull/367

### Framework

1. Add missing check in `update_auth_key_and_originating_address_table` to abort
   if current and new authentication key are identical.
1. Add missing check in `update_auth_key_and_originating_address_table` to abort
   if new authentication key is already in the `OriginatingAddress` table.
1. Fix backwards doc comments on `RotationProofChallenge`, adding missing `/`.
1. Add `set_originating_address` reconciliation function.
1. Add `originating_address()` view function.
1. Add missing `/` to assorted error codes.

Closes #13517

### CLI

1. Add Ledger support to key rotation subcommand.
1. Abstract, consolidate key rotation command options.
1. Add profile subcommands to delete or rename a CLI profile, and to expose its
   private key.
1. Update tests.
1. Use arg group instead of interactive prompt for new profile name since the
   latter did not prevent errors on subsequent re-attempts.
1. Simplify and clarify BIP44-related command annotations where appropriate.

Closes #13515

## Verification

Note that this requires the Ledger Aptos App >=0.6.9, which contains the
required blind signing capabilities that previously blocked this PR. The version
can be installed via Ledger Live.

Commands tested manually, via CLI compiled from source:

1.  From in `crates/aptos`:

```sh
git ls-files | entr -c cargo build
```

1. Update path with environment variable for `aptos` binary:

```sh
export PATH="wherever-you-have-aptos-core/aptos-core/target/debug:$PATH"
```

1. Verify which `aptos`:

```sh
which aptos
```

1. Run the tutorials at https://github.com/aptos-labs/developer-docs/pull/367

## Additional local checks

The following commands were run in the below directories:

In `aptos-core/`:

```sh
cargo build -p aptos-cached-packages
pre-commit run --all-files
scripts/rust_lint.sh
```

In `aptos-core/crates/aptos/`

```sh
cargo test
```

In `aptos-core/aptos-move/framework/aptos-framework/`

```sh
aptos move test --dev
```
